src/ops/indexing.rs

```diff
@@ -0,0 +1,1799 @@
+//!
+//! # Multi-axes indexing
+//!
+//! Multi-axes indexing with combinations of the above operations is also supported by combining the
+//! operations in a tuple with the restriction that `Ellipsis` can only be used once.
+//!
+//! ## Examples
+//!
+//! ```rust
+//! // See the multi-dimensional example code for mlx python https://ml-explore.github.io/mlx/build/html/usage/indexing.html
+//!
+//! use mlx::prelude::*;
+//!
+//! let a = Array::from_iter(0..8, &[2, 2, 2]);
+//!
+//! // a[:, :, 0]
+//! let mut s1 = a.index((.., .., 0));
+//!
+//! let expected = Array::from_slice(&[0, 2, 4, 6], &[2, 2]);
+//! // TODO: assert_eq!(s1, expected);
+//!
+//! // a[..., 0]
+//! let mut s2 = a.index((Ellipsis, 0));
+//!
+//! let expected = Array::from_slice(&[0, 2, 4, 6], &[2, 2]);
+//! // TODO: assert_eq!(s1, expected);
+//! ```
+
+use std::{
+    borrow::Cow,
+    ops::{Bound, RangeBounds},
+    rc::Rc,
+};
+
+use mlx_macros::default_device;
+use smallvec::{smallvec, SmallVec};
+
+use crate::{
+    error::{InvalidAxisError, SliceError, TakeAlongAxisError, TakeError},
+    utils::resolve_index_unchecked,
+    Array, StreamOrDevice,
+};
+
+/* -------------------------------------------------------------------------- */
+/*                                Custom types                                */
+/* -------------------------------------------------------------------------- */
+
+#[derive(Debug, Clone, Copy)]
+pub struct NewAxis;
+
+#[derive(Debug, Clone, Copy)]
+pub struct Ellipsis;
+
+#[derive(Debug, Clone, Copy)]
+pub struct StrideBy<I> {
+    pub inner: I,
+    pub stride: i32,
+}
+
+pub trait IntoStrideBy: Sized {
+    fn stride_by(self, stride: i32) -> StrideBy<Self>;
+}
+
+impl<T> IntoStrideBy for T {
+    fn stride_by(self, stride: i32) -> StrideBy<Self> {
+        StrideBy {
+            inner: self,
+            stride,
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct RangeIndex {
+    start: Bound<i32>,
+    stop: Bound<i32>,
+    stride: i32,
+}
+
+impl RangeIndex {
+    pub(crate) fn new(start: Bound<i32>, stop: Bound<i32>, stride: Option<i32>) -> Self {
+        // let start = match start {
+        //     Bound::Included(start) => start,
+        //     Bound::Excluded(start) => start + 1,
+        //     Bound::Unbounded => 0,
+        // };
+        // let stop = match stop {
+        //     Bound::Included(stop) => Some(stop + 1),
+        //     Bound::Excluded(stop) => Some(stop),
+        //     Bound::Unbounded => None,
+        // };
+        let stride = stride.unwrap_or(1);
+        Self {
+            start,
+            stop,
+            stride,
+        }
+    }
+
+    pub(crate) fn stride(&self) -> i32 {
+        self.stride
+    }
+
+    pub(crate) fn start(&self, size: i32) -> i32 {
+        // _start ?? (stride < 0 ? size - 1 : 0)
+
+        // if self.stride.is_negative() {
+        //     self.start + size
+        // } else {
+        //     self.start
+        // }
```

Comment:
Looks like all these methods have some extra logic comments?

Reply:
Addressed in 55ca17d931a3dff70b189ecb3086957543ca4d55